### PR TITLE
Add cmo purchase new module

### PR DESCRIPTION
**Pull request by newtratip**
P'Nuii please review my code

### CODE REVIEW

File: specific/cmo_purchase/models/product.py [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+            args = [('id', 'in', [])]
+
+        # Search products by category
+        if context.get('po_type_id', False):
+            po_type_id = context.get('po_type_id', [])
+            PoTypeConfig = self.env['purchase.order.type.config']
+            po_type_config = PoTypeConfig.browse(po_type_id)
+            categ_ids = po_type_config.category_id.mapped('id')
+            product = self.search([('categ_id', 'in', categ_ids)])
+            product_ids = product.mapped('id')
```
Comment (kittiu) [inline specific/cmo_purchase/models/product.py]: You can use `product.ids`

File: specific/cmo_purchase/models/product.py [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+            args = [('id', 'in', product_ids)] + args
+        elif 'order_ref' in context:
+            args = [('id', 'in', [])]
+
+        # Search products by category
+        if context.get('po_type_id', False):
+            po_type_id = context.get('po_type_id', [])
+            PoTypeConfig = self.env['purchase.order.type.config']
+            po_type_config = PoTypeConfig.browse(po_type_id)
+            categ_ids = po_type_config.category_id.mapped('id')
```
Comment (kittiu) [inline specific/cmo_purchase/models/product.py]: you can use po_type_config.category_id.ids

File: specific/cmo_purchase/models/purchase.py [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+    )
+
+    @api.onchange('po_type_id')
+    def _onchange_po_type_id(self):
+        self.project_id = False
+        self.order_ref = False
+        self.event_date_description = False
+        self.venue_description = False
+        self.order_line = False
+        self.invoice_method = self.po_type_id and \
```
Comment (kittiu) [inline specific/cmo_purchase/models/purchase.py]: In API v8, if we sure the filed is many2one, we con't need to check for False.
Here is enough,
`... = self.po_type_id.invoice_method or False`

File: specific/cmo_purchase/models/purchase.py [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+
+    @api.multi
+    def _check_amount_untaxed(self):
+        for order in self:
+            po_project = order.po_type_id and order.po_type_id.po_project \
+                or False
+            remaining_cost = order.project_id and \
+                order.project_id.remaining_cost or False
+            if po_project and remaining_cost is not False:
+                if order.amount_untaxed > remaining_cost:
```
Comment (kittiu) [inline specific/cmo_purchase/models/purchase.py]: No blocking, but it is always better use `float_compare` for float comparison.

File: specific/cmo_purchase/models/purchase.py [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+    @api.multi
+    def write(self, vals):
+        res = super(PurchaseOrder, self).write(vals)
+        self._check_amount_untaxed()
+        return res
+
+    @api.multi
+    def _check_amount_untaxed(self):
+        for order in self:
+            po_project = order.po_type_id and order.po_type_id.po_project \
```
Comment (kittiu) [inline specific/cmo_purchase/models/purchase.py]: `...= order.po_type_id.po_project or False` is enough

File: specific/cmo_purchase/models/purchase.py [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+                or False
+            remaining_cost = order.project_id and \
```
Comment (kittiu) [inline specific/cmo_purchase/models/purchase.py]: This one too, can make is shorter.

File: specific/cmo_purchase/models/purchase.py [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+                order.project_id.remaining_cost or False
+            if po_project and remaining_cost is not False:
+                if order.amount_untaxed > remaining_cost:
+                    raise ValidationError(
```
Comment (kittiu) [inline specific/cmo_purchase/models/purchase.py]: Use translation function for all the messgages, i..e,
`raise ValidationError(_("xxxxx"))`

File: specific/cmo_purchase/models/purchase.py [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+    _inherit = 'purchase.order.line'
+
+    product_ref = fields.Many2one(
+        'product.product',
+        string='Product Ref',
+    )
+    date_planned = fields.Date(
+        required=False,
+    )
+    group_id = fields.Many2one(
```
Comment (kittiu) [inline specific/cmo_purchase/models/purchase.py:112]: No label, right?

**Comment by kittiu** [comment]
Done review, please fix.

File: specific/cmo_purchase/models/purchase.py (hunk 8 above)
Comment (newtratip) [reply specific/cmo_purchase/models/purchase.py:112]: yes, not label.
I'm inherit field from core module which declare require is False only.

Comment (kittiu) [reply specific/cmo_purchase/models/purchase.py:112]: OK

**Pull request merged**